discretize/BaseMesh.py

```diff
@@ -1,41 +1,57 @@
 import numpy as np
+import properties
+import os
+import json
 from discretize import utils
 
 
-class BaseMesh(object):
-    """BaseMesh does all the counting you don't want to do.
+class BaseMesh(properties.HasProperties):
+    """
+    BaseMesh does all the counting you don't want to do.
     BaseMesh should be inherited by meshes with a regular structure.
-
-    :param numpy.array n: (or list) number of cells in each direction (dim, )
-    :param numpy.array x0: (or list) Origin of the mesh (dim, )
-
     """
 
+    # Properties
+    n = properties.Array(
+        "number of cells in each direction (dim, )",
+        dtype=int,
+        required=True,
+        shape=('*',)
+    )
+
+    x0 = properties.Array(
+        "origin of the mesh (dim, )",
+        dtype=float,
+        shape=('*',),
+        required=True
+    )
+
+    # Instantiate the class
     def __init__(self, n, x0=None):
+        self.n = n  # number of dimensions
 
-        # Check inputs
         if x0 is None:
-            x0 = np.zeros(len(n))
-
-        if not len(n) == len(x0):
-            raise Exception("Dimension mismatch. x0 != len(n)")
-
-        if len(n) > 3:
-            raise Exception("Dimensions higher than 3 are not supported.")
-
-        # Ensure x0 & n are 1D vectors
-        self._n = np.array(n, dtype=int).ravel()
-        self._x0 = np.array(x0, dtype=float).ravel()
-        self._dim = len(self._x0)
-
-    @property
-    def x0(self):
-        """Origin of the mesh
-
-        :rtype: numpy.array
-        :return: x0, (dim, )
-        """
-        return self._x0
+            self.x0 = np.zeros(len(self.n))
+        else:
+            self.x0 = x0
+
+    # Validators
+    @properties.validator('n')
+    def check_n_shape(self, change):
+        change['value'] = np.array(change['value'], dtype=int).ravel()
+        if len(change['value']) > 3:
+            raise Exception(
+                "Dimensions of {}, which is higher than 3 are not "
+                "supported".format(change['value'])
+            )
+
+    @properties.validator('x0')
+    def check_x0_vs_n(self, change):
```

Comment:
A couple picky thoughts on this:
- `self.n` _could_ be `None` if someone deletes it or sets it to undefined. `len(None)` would give an ugly error. Possibly worth adding an extra sanity check for that?
- Also, possibly worth running this check in the other direction (ie you validate `x0` vs `n`, but not `n` vs `x0`) - may come up depending on how someone is using this.